testing/trino-testing-services/src/main/java/io/trino/testng/services/ReportResourceHungryTests.java

```diff
@@ -0,0 +1,163 @@
+                    testClass,
+                    getStackTraceAsString(e));
+        }
+    }
+
+    @Override
+    public void onAfterClass(ITestClass testClass)
+    {
+        try {
+            reportResources(testClass);
+        }
+        catch (ReflectiveOperationException | Error e) {
+            reportListenerFailure(
+                    ReportResourceHungryTests.class,
+                    "Failed to process %s: \n%s",
+                    testClass,
+                    getStackTraceAsString(e));
+        }
+    }
+
+    private void reportResources(ITestClass testClass)
+            throws ReflectiveOperationException
+    {
+        for (Object instance : testClass.getInstances(false)) {
+            for (Class<?> clazz = instance.getClass(); clazz != null; clazz = clazz.getSuperclass()) {
+                for (Field field : clazz.getDeclaredFields()) {
+                    field.setAccessible(true);
+                    Object value = field.get(instance);
+                    if (value == null) {
+                        continue;
+                    }
+                    for (Rule rule : rules) {
+                        if (rule.isResourceful(field, value)) {
+                            reportListenerFailure(
+                                    ReportResourceHungryTests.class,
+                                    "Field %s of %s [%s] is set (to %s) before class is started, or after class is finished. \n" +
+                                            "Resources must not be allocated in field initializer or test constructor, and must be freed when test class is completed. \n" +
+                                            "Depending which rule has been violated, if the reported class holds on to resources only conditionally, you can add @ResourcePresence to declare that.",
+                                    field,
+                                    instance,
+                                    testClass.getRealClass(),
+                                    value);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private static Optional<Rule> isInstance(String className)
+    {
+        return tryLoadClass(className)
+                .map(clazz -> (field, value) -> clazz.isInstance(value));
+    }
+
+    private static Rule declaredClassPattern(@Language("RegExp") String classNamePattern)
+    {
+        Pattern pattern = Pattern.compile(classNamePattern);
+        return (field, value) -> pattern.matcher(field.getType().getName()).matches();
+    }
+
+    private static Optional<Class<?>> tryLoadClass(String className)
+    {
+        try {
+            return Optional.of(Thread.currentThread().getContextClassLoader().loadClass(className));
+        }
+        catch (ClassNotFoundException e) {
+            return Optional.empty();
+        }
+    }
+
+    private interface Rule
+    {
+        boolean isResourceful(Field field, Object value);
```

Comment:
nit: resourceful means "having the ability to find quick and clever ways to overcome difficulties", how about `isExpensiveResource`?

Reply:
good idea